### PR TITLE
Trace `Runner.run_streamed()` in OpenAI Agents SDK autolog

### LINKED ISSUE #22739: [FR] Support tracing of Runner.run_streamed() in OpenAI Agents SDK autolog
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

`mlflow.openai.autolog()` only patches `AgentRunner.run`, not `AgentRunner.run_streamed`. As a result, traces from `Runner.run_streamed()` have an empty root span (`inputs=""`, `outputs=""`). This FR adds a matching patch so streamed runs reach feature parity with `Runner.run()`.

### Motivation

> #### What is the use case for this feature?

Streaming agent workflows (token-by-token UIs, SSE endpoints) use `Runner.run_streamed()` instead of `Runner.run()`. Users enable `mlflow.openai.autolog()` and expect the streamed trace to match the non-streamed one.

> #### Why is this use case valuable to support for MLflow users in general?

Streaming is a common pattern for LLM applications. When the root span has no `inputs` / `outputs`, the trace list UI shows empty `Request` / `Response` cells, and searching traces by request/response text does not match streamed runs. The non-streamed path already works via `_patched_agent_run` — the streamed path should reach parity.

> #### Why is this use case valuable to support for your project(s) or organization?

We hit this while evaluating `mlflow.openai.autolog()` for use with the OpenAI Agents SDK on a self-hosted MLflow deployment. Without this feature, autolog is effectively unusable for any code path that chooses the streaming API.

> #### Why is it currently difficult to achieve this use case?

There is no user-side workaround: for streamed runs, the root span is created by the autolog tracer callback (`_agent_tracer.py`) with hard-coded empty strings. Wrapping `Runner.run_streamed()` manually to fill the root span would duplicate autolog's responsibilities.

### Reproduction

Environment: self-hosted MLflow 3.11.2.dev0, `openai-agents==0.13.5`, `openai==2.30.0`, OpenAI Responses API.

```python
import asyncio
import mlflow
from agents import Agent, Runner

mlflow.openai.autolog()
mlflow.set_experiment("repro-run-streamed")

agent = Agent(name="assistant", instructions="You are a helpful assistant.")


async def main():
    await Runner.run(agent, "Hello")  # OK

    streamed = Runner.run_streamed(agent, "Hello")  # Broken
    async for _ in streamed.stream_events():
        pass


asyncio.run(main())
```

| Field | `Runner.run` | `Runner.run_streamed` |
|---|---|---|
| Root span name | `AgentRunner.run` | `Agent workflow` |
| Root `inputs` | `'Hello'` | `''` |
| Root `outputs` | `'Hello! How can I help you today?'` | `''` |
| `request_preview` / `response_preview` | populated | `'""'` / `'""'` |

### Root cause

- `mlflow/openai/autolog.py:89` patches only `AgentRunner.run`.
- Without a patch, the root span falls back to `MlflowOpenAgentTracingProcessor` in `_agent_tracer.py`, which hard-codes `inputs=""` at L106 and `outputs=""` at L123.

### Proposed implementation (sketch, not yet tested end-to-end)

Add a `safe_patch` for `AgentRunner.run_streamed` that mirrors `_patched_agent_run`, adapted for streaming semantics:

- `run_streamed()` is synchronous: it calls `asyncio.create_task(start_streaming(...))` and returns a `RunResultStreaming` immediately (confirmed against `openai-agents==0.13.5`). The span must therefore be opened in the patched method and closed from a wrapped `stream_events()` async generator, not from the patched method's body.
- `mlflow.tracing.fluent.start_span` is a `@contextlib.contextmanager` and cannot span the sync return. Use `start_span_no_context` + `set_span_in_context` / `detach_span_from_context` (the same pattern already used elsewhere in `_agent_tracer.py`).
- Hypothesis: if the new patch names its span `"AgentRunner.run_streamed"`, the name check in `MlflowOpenAgentTracingProcessor.on_trace_start` (`active_span.name == _AGENT_RUN_SPAN_NAME`) must be widened (e.g. to a set containing both names) so the tracer callback reuses the patched span instead of creating a duplicate root. I have not validated this interaction with a runnable patch yet.
- Close with `ERROR` status on exceptions; guard against double-close if `stream_events()` is iterated more than once.

Sketch (conceptual only):

```python
_AGENT_RUN_SPAN_NAMES = {"AgentRunner.run", "AgentRunner.run_streamed"}


def _patched_agent_run_streamed(original, self, *args, **kwargs):
    inputs = construct_full_inputs(original, self, *args, **kwargs)
    attributes = {
        k: v for k, v in inputs.items()
        if k not in ("starting_agent", "input", "run_config")
    }

    span = start_span_no_context(
        name="AgentRunner.run_streamed",
        span_type=SpanType.AGENT,
        inputs=inputs.get("input"),
        attributes=attributes,
    )
    token = set_span_in_context(span)

    try:
        result = original(self, *args, **kwargs)
    except Exception:
        detach_span_from_context(token)
        span.end(status=SpanStatusCode.ERROR)
        raise

    _original_stream_events = result.stream_events
    ended = False

    async def _wrapped_stream_events():
        nonlocal ended
        status = SpanStatusCode.OK
        try:
            async for event in _original_stream_events():
                yield event
            span.set_outputs(result.final_output)
        except Exception:
            status = SpanStatusCode.ERROR
            raise
        finally:
            if not ended:
                ended = True
                detach_span_from_context(token)
                span.end(status=status)

    result.stream_events = _wrapped_stream_events
    return result
```

### Related

- #17227 (merged): introduced `_patched_agent_run` for `AgentRunner.run`. This FR is the `run_streamed` counterpart.

### What machine learning domain(s) is this feature request about?

- [x] `domain/genai`: LLMs, Agents, and other GenAI-related use cases
- [ ] `domain/classical-ml`: Traditional machine learning, such as linear regression.
- [ ] `domain/deep-learning`: Deep learning and neural networks.
- [ ] `domain/platform`: MLflow platform foundation, not specific to a particular machine learning domain.

### What area(s) of MLflow is this feature request about?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by ktrk115**
### Related Issues/PRs

Closes #22739

### What changes are proposed in this pull request?

`mlflow.openai.autolog()` previously patched only `AgentRunner.run`, so traces from `Runner.run_streamed()` had an empty root span (`inputs=""`, `outputs=""`). This PR adds a matching patch so streamed runs reach feature parity with non-streamed runs.

`run_streamed()` is synchronous and returns a `RunResultStreaming` immediately while running the work as an `asyncio` background task — events are produced as the user consumes `result.stream_events()`. The new `_patched_agent_run_streamed`:

1. Opens a root span (`AgentRunner.run_streamed`, span_type=`AGENT`) without a context manager.
2. Attaches it to the OTel context **before** invoking `original`, so the background task inherits it as parent.
3. Wraps `result.stream_events()` so the span is detached + ended when iteration completes (or on exception).

To decouple root-span detection from the human-readable name, `MlflowOpenAgentTracingProcessor.on_trace_start` now reuses the active span via a private marker attribute (`mlflow.openai.agent_run_root`) instead of comparing `span.name` against `"AgentRunner.run"`. A small helper `_build_agent_run_span_args` shares the inputs/attributes construction (including the `run_config` exclusion comment from #19911) between `_patched_agent_run` and `_patched_agent_run_streamed`.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

New tests in `tests/openai/test_openai_agent_autolog.py`:

- `test_autolog_agent_run_streamed` — mocks `OpenAIResponsesModel.stream_response` to yield `ResponseTextDeltaEvent` + `ResponseCompletedEvent`, runs `Runner.run_streamed`, and asserts the root span has `name == "AgentRunner.run_streamed"`, span_type AGENT, and the expected `inputs`/`outputs` propagated to `request_preview`/`response_preview`. Verifies all child spans descend from the streamed root.
- `test_autolog_agent_run_streamed_exception` — patches `stream_response` to raise during iteration, asserts the trace ends in ERROR status with an `exception` event on the root span.

Manual verification against a real Azure OpenAI deployment (`gpt-5-mini` via Chat Completions) using a local MLflow server confirms:

| Trace | Root span name | Root inputs | Root outputs |
|---|---|---|---|
| `Runner.run()` | `AgentRunner.run` | `"Hello"` | `"Hi — how can I help you today?"` |
| **`Runner.run_streamed()`** | **`AgentRunner.run_streamed`** | **`"Hello again"`** | **`"Hi — how can I help you today?"`** |

Both traces show the expected values in the MLflow UI's Traces list and trace-detail panel.

image

**Note:** Token usage is not recorded for streamed runs because the SDK's child generation spans (`Generation` / `AsyncCompletions`) don't receive usage info from the streaming response. This is out of scope for this PR; I'll file a separate issue for it.

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. `mlflow.openai.autolog()` now traces `Runner.run_streamed()` in the OpenAI Agents SDK with the same fidelity as `Runner.run()`, populating the root span's inputs/outputs and the trace request/response previews instead of leaving them empty.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Minor releases are expected to contain larger changes, such as new features and improvements. Non-critical bug fixes and doc updates can be included as well. By default, your PR should target the next minor release.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Patch releases are typically only performed when there has been a major regression or bug in the latest release. For the sake of stability, your PR should not be included in a patch release unless it is a critical fix, or if the risk level of your PR is exceedingly low.

</details>

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by ktrk115** [comment]
@copilot start review

File: mlflow/openai/_agent_tracer.py [hunk 1]
```diff
@@ -282,24 +286,104 @@ def _parse_response_span_data(span_data: oai.ResponseSpanData) -> tuple[Any, Any
...
+    span = start_span_no_context(
+        name=_AGENT_RUN_STREAMED_SPAN_NAME,
+        span_type=SpanType.AGENT,
+        inputs=inputs.get("input"),
+        attributes=attributes,
+    )
+    span.set_attribute(_AGENT_RUN_ROOT_MARKER, True)
+    # Attach the span before original() runs so the asyncio task it spawns
+    # captures this span as its OTel context.
+    token = set_span_in_context(span)
```
Comment (kriscon-db) [inline mlflow/openai/_agent_tracer.py:350]: `_patched_agent_run_streamed` attaches an OTel context token and span unconditionally, but `_finalize_streamed_span` only fires from the wrapped `result.stream_events`. If a caller does `result = Runner.run_streamed(agent, "x")` and discards `result` without iterating, the wrapped function never runs and the span and OTel token leak. MLflow already uses `weakref.finalize` for this exact pattern in autologging (see `mlflow/sklearn/__init__.py:938` and `mlflow/pyspark/ml/__init__.py:678`, both of which attach a finalizer to a user-facing result object so cleanup fires on GC). Applying it here covers the discarded-result case while still allowing explicit iteration to clean up via `finalizer.detach()` in the wrapped generator's `finally`:

```python
finalizer = weakref.finalize(result, _finalize_streamed_span, span, token)

async def wrapped_stream_events(*args, **kwargs):
    error = None
    try:
        async for event in original_stream_events(*args, **kwargs):
            yield event
    except Exception as e:
        error = e
        raise
    finally:
        finalizer.detach()  # cancel GC finalizer; explicit cleanup runs
        outputs = None if error else getattr(result, "final_output", None)
        _finalize_streamed_span(span, token, error=error, outputs=outputs)
```

-- Claude-drafted, reviewed before posting.

File: mlflow/openai/_agent_tracer.py [hunk 2]
```diff
@@ -282,24 +286,104 @@ def _parse_response_span_data(span_data: oai.ResponseSpanData) -> tuple[Any, Any
...
+        detach_span_from_context(token)
+        if error is not None:
+            span.record_exception(error)
+            span.end()
+            return
+        if outputs is not None:
+            span.set_outputs(outputs)
+        span.end(status=SpanStatusCode.OK)
+    except Exception:
+        _logger.debug("Failed to finalize streamed agent run span", exc_info=True)
```
Comment (kriscon-db) [inline mlflow/openai/_agent_tracer.py]: `_finalize_streamed_span` swallows any exception during cleanup at `_logger.debug(...)`, which is off by default. If `span.end()` or `detach_span_from_context()` ever fails (e.g., span store unreachable, OTel context already torn down by another path), the span is left unclosed and the operator has no signal. Per project convention, implementation-detail goes to `_logger.debug` but operationally-relevant failures go to `_logger.warning`; a leaked or stale span counts as operationally relevant.

```suggestion
        _logger.warning("Failed to finalize streamed agent run span", exc_info=True)
```

-- Claude-drafted, reviewed before posting.

**Comment by kriscon-db** [comment]
Can you also post a clear before and after set of screenshots?

File: mlflow/openai/_agent_tracer.py (hunk 1 above)
Comment (ktrk115) [reply mlflow/openai/_agent_tracer.py:350]: Thanks for the careful catch!

Addressed in https://github.com/mlflow/mlflow/pull/22962/commits/ea5f18a5633fc563a42f3ba3c320177ccaaf078c (+ test for the discarded-result case).
Two non-obvious deviations from the suggested snippet:

- Capturing `result.stream_events` as a bound method puts `result` in the closure cell and defeats the GC finalizer, so the wrapped function takes `type(result).stream_events` (unbound) plus a `weakref.ref(result)`.
- Split `_finalize_streamed_span` into `_safe_detach_span_context` + the end logic. Otherwise a cross-context `detach` from the finalizer (`ValueError`) skips `span.end()` and the span leaks anyway. Same `except ValueError` idiom as `mlflow/langchain/langchain_tracer.py`. https://github.com/mlflow/mlflow/blob/093c39a4fe0be51e0ffd9fd8355da2225b50032a/mlflow/langchain/langchain_tracer.py#L273-L285

-- Claude-drafted, revised before posting.

File: mlflow/openai/_agent_tracer.py (hunk 2 above)
Comment (ktrk115) [reply mlflow/openai/_agent_tracer.py]: Thanks, addressed in https://github.com/mlflow/mlflow/pull/22962/commits/ea5f18a5633fc563a42f3ba3c320177ccaaf078c.

**Comment by ktrk115** [comment]
> Can you also post a clear before and after set of screenshots?

Captured against Azure OpenAI (`gpt-5-mini`, Chat Completions). Each experiment runs `Runner.run("Hello from run")` and `Runner.run_streamed("Hello from run_streamed")`.

| Before (master) | After (this PR) |
|---|---|
| image | image|

The `run_streamed` row has `null` Request and `""` Response on master; with this PR it populates them just like `run`.

> Note: the `run_streamed` row's `Tokens` column is still empty after this PR. It's a separate problem — the OpenAI Agents SDK's streaming path doesn't surface usage on the child generation spans, which MLflow aggregates from. I'll file a follow-up issue for that.

**Comment by kriscon-db** [comment]
Hi @ktrk115!  Thanks for the submission.  This LGTM.  Can you address the breaking tests and ill rerun the CI and then stamp and merge this.

**Comment by ktrk115** [comment]
Hi @kriscon-db, thanks for the review!

> MLflow tests / windows (3) (pull_request)Failing after 46m

The failure is in `tests/webhooks/test_e2e.py` (`Failed to start server at http://localhost:52127/health`), which this PR doesn't touch, and it looks like a flaky test. Could you rerun it when you have a moment?

**Review by kriscon-db — APPROVED** [review]
